compiler/rustc_hir_analysis/src/check/intrinsicck.rs

```diff
@@ -84,34 +106,16 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
             ty::Adt(adt, substs) if adt.repr().simd() => {
                 let fields = &adt.non_enum_variant().fields;
                 let elem_ty = fields[FieldIdx::from_u32(0)].ty(self.tcx, substs);
+
                 match elem_ty.kind() {
-                    ty::Never | ty::Error(_) => return None,
-                    ty::Int(IntTy::I8) | ty::Uint(UintTy::U8) => {
-                        Some(InlineAsmType::VecI8(fields.len() as u64))
-                    }
-                    ty::Int(IntTy::I16) | ty::Uint(UintTy::U16) => {
-                        Some(InlineAsmType::VecI16(fields.len() as u64))
-                    }
-                    ty::Int(IntTy::I32) | ty::Uint(UintTy::U32) => {
-                        Some(InlineAsmType::VecI32(fields.len() as u64))
-                    }
-                    ty::Int(IntTy::I64) | ty::Uint(UintTy::U64) => {
-                        Some(InlineAsmType::VecI64(fields.len() as u64))
-                    }
-                    ty::Int(IntTy::I128) | ty::Uint(UintTy::U128) => {
-                        Some(InlineAsmType::VecI128(fields.len() as u64))
-                    }
-                    ty::Int(IntTy::Isize) | ty::Uint(UintTy::Usize) => {
-                        Some(match self.tcx.sess.target.pointer_width {
-                            16 => InlineAsmType::VecI16(fields.len() as u64),
-                            32 => InlineAsmType::VecI32(fields.len() as u64),
-                            64 => InlineAsmType::VecI64(fields.len() as u64),
-                            _ => unreachable!(),
-                        })
+                    ty::Array(ty, len) => {
+                        let Some(len) = len.try_eval_target_usize(self.tcx, self.tcx.param_env(adt.did())) else {
+                            bug!("un-evaled const in SIMD type {ty} {len}");
```

Comment:
Is this OK? I think so but I'm not 100% sure.

Reply:
So, please use `span_bug!` at least.

Reply:
Also, we shouldn't really be using even `span_bug!`. This will trigger an ICE, not a diagnostic. We should not be triggering an ICE just because we failed to get a value out of the to-be-evaluated constant parameter. We should trigger a diagnostic indicating the failure to evaluate and explain why.

Reply:
Oh wait, I'll just return `None` here. It matches what happens with the check for fields on `repr(simd)`.

Reply:
I'm worried that this may be triggered with some creative usage of the `generic_const_exprs` feature. We should probably be just emitting a regular diagnostic here?

Reply:
Basically, const parameters fail to evaluate for bullshit reasons that I don't personally understand all the time, but I do know they are going to happen on the regular and that even if this const eval path is hidden by some eager evaluation before this right _now_, the likelihood this code path will be exposed by a later code change is approximately 99.999% (something something "lazy normalization" probably), so we should just assume that it's possible for this to fail and default to assuming we should be giving a reasonable diagnostic.

Reply:
race condition

Reply:
Yeah, I've fixed it.